pyface/ui/qt4/data_view/data_view_widget.py

```diff
@@ -21,29 +25,162 @@
+        return pyface_selection_modes[qt_selection_mode]
+
+    def _set_control_selection_mode(self, selection_mode):
+        """ Toolkit specific method to change the selection mode. """
+        qt_selection_mode = qt_selection_modes[selection_mode]
+        self.control.setSelectionMode(qt_selection_mode)
+        self.selection = []
+
+    def _get_control_selection(self):
+        """ Toolkit specific method to get the selection. """
+        if self.selection_type == 'row':
+            selection = []
+            for index in self.control.selectedIndexes():
+                row = self._item_model._to_row_index(index)
+                if (row, ()) not in selection:
+                    selection.append((row, ()))
+        elif self.selection_type == 'column':
+            selection = []
+            for index in self.control.selectedIndexes():
+                row = self._item_model._to_row_index(index)[:-1]
+                column = self._item_model._to_column_index(index)
+                if (row, column) not in selection:
+                    selection.append((row, column))
+        else:
+            selection = [
+                (
+                    self._item_model._to_row_index(index),
+                    self._item_model._to_column_index(index),
+                )
+                for index in self.control.selectedIndexes()
+            ]
+        return selection
+
+    def _set_control_selection(self, selection):
+        """ Toolkit specific method to change the selection. """
+        selection_model = self.control.selectionModel()
+        select_flags = QItemSelectionModel.Select
+        qt_selection = QItemSelection()
+
+        if self.selection_type == 'row':
+            select_flags |= QItemSelectionModel.Rows
+            for row, column in selection:
+                index = self._item_model._to_model_index(row, (0,))
+                qt_selection.select(index, index)
+        elif self.selection_type == 'column':
+            select_flags |= QItemSelectionModel.Columns
+            for row, column in selection:
+                index = self._item_model._to_model_index(
+                    row + (0,), column)
+                qt_selection.select(index, index)
+        else:
+            for row, column in selection:
+                index = self._item_model._to_model_index(row, column)
+                qt_selection.select(index, index)
+        selection_model.clearSelection()
+        selection_model.select(qt_selection, select_flags)
+
+    def _observe_control_selection(self, remove=False):
+        selection_model = self.control.selectionModel()
+        if remove:
+            try:
+                selection_model.selectionChanged.disconnect(
+                  self._update_selection
+                )
+            except (TypeError, RuntimeError):
+                # has already been disconnected
+                logger.info("selectionChanged already disconnected")
```

Comment:
I think it's because the selection model is its own Qt QObject and has its own lifecycle which we don't control.  So it may be finalised first during tear-down and as a result the signals are disconnected.  We don't want to take any sort of ownership of the selection model if we can avoid it - it would add a _lot_ of complexity vs. just acknowledging that Qt might beat us to disconnecting things.

It's fine if Qt disconnects first: the reason we are doing this explicitly here is to prevent us firing trait events when the widget is being torn down, so it's fine to already be in a safe state.

More generally, the explicit disconnection is belt-and-suspenders on our part.  In many cases we could just let everything be finalized by Qt and all the Qt-side stuff will be disconnected properly and any extra trait change events are likely harmless (particularly if handlers explicitly check for the status of the control).